coordinator/src/main/java/org/apache/uniffle/coordinator/HealthSelectStorageStrategy.java

```diff
@@ -0,0 +1,266 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.common.RemoteStorageInfo;
+import org.apache.uniffle.common.filesystem.HadoopFilesystemProvider;
+import org.apache.uniffle.common.util.ThreadUtils;
+
+/**
+ * HealthSelectStorageStrategy considers that when allocating apps to different remote paths,
+ * remote paths that can write and read. Therefore, it may occur that all apps are written to the same cluster.
+ * At the same time, if a cluster has read and write exceptions, we will automatically avoid the cluster.
+ */
+public class HealthSelectStorageStrategy implements SelectStorageStrategy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HealthSelectStorageStrategy.class);
+  /**
+   * store appId -> remote path to make sure all shuffle data of the same application
+   * will be written to the same remote storage
+   */
+  private final Map<String, RemoteStorageInfo> appIdToRemoteStorageInfo;
+  /**
+   * store remote path -> application count for assignment strategy
+   */
+  private final Map<String, RankValue> remoteStoragePathRankValue;
+  private final Map<String, RemoteStorageInfo> availableRemoteStorageInfo;
+  private List<Map.Entry<String, RankValue>> sizeList;
+  private FileSystem fs;
+  private Configuration conf;
+  private final int fileSize;
+  private final int readAndWriteTimes;
+
+  public HealthSelectStorageStrategy(CoordinatorConf cf) {
+    conf = new Configuration();
+    fileSize = cf.getInteger(CoordinatorConf.COORDINATOR_REMOTE_STORAGE_FILE_SIZE);
+    readAndWriteTimes = cf.getInteger(CoordinatorConf.COORDINATOR_REMOTE_STORAGE_ACCESS_TIMES);
+    this.appIdToRemoteStorageInfo = Maps.newConcurrentMap();
+    this.remoteStoragePathRankValue = Maps.newConcurrentMap();
+    this.availableRemoteStorageInfo = Maps.newHashMap();
+    this.sizeList = Lists.newCopyOnWriteArrayList();
+    ScheduledExecutorService readWriteRankScheduler = Executors.newSingleThreadScheduledExecutor(
+        ThreadUtils.getThreadFactory("readWriteRankScheduler-%d"));
+    // should init later than the refreshRemoteStorage init
+    readWriteRankScheduler.scheduleAtFixedRate(this::checkReadAndWrite, 1000,
+        cf.getLong(CoordinatorConf.COORDINATOR_REMOTE_STORAGE_HEALTH_SCHEDULE_TIME), TimeUnit.MILLISECONDS);
+  }
+
+  public void checkReadAndWrite() {
+    if (remoteStoragePathRankValue.size() > 1) {
+      for (String path : remoteStoragePathRankValue.keySet()) {
+        Path remotePath = new Path(path);
+        Path testPath = new Path(path + "/rssTest");
+        long startWriteTime = System.currentTimeMillis();
+        try {
+          fs = HadoopFilesystemProvider.getFilesystem(remotePath, conf);
+          for (int j = 0; j < readAndWriteTimes; j++) {
```

Comment:
If we write the same file, the file will be allocated the same DN, our sample  will only test only one DN.

Reply:
![Uploading image.png…]()
Although we are writing the same file, the stream of this file has been closed, and a batch of new DN should be reallocated next time we write.

Reply:
![image](https://user-images.githubusercontent.com/84573424/188814033-d61b5820-db6e-40cc-824d-f2e46b03268c.png)

Reply:
![image](https://user-images.githubusercontent.com/84573424/188815077-11ef41e0-9b77-4569-bafc-93bc7319fa7e.png)

Reply:
OK.